internal/xds/resolver/xds_http_filters_test.go

```diff
@@ -0,0 +1,678 @@
+	ctx context.Context
+}
+
+func (cs *clientStream) Context() context.Context {
+	return cs.ctx
+}
+
+func newHTTPFilter(t *testing.T, name, typeURL, path, err string) *v3httppb.HttpFilter {
+	return &v3httppb.HttpFilter{
+		Name: name,
+		ConfigType: &v3httppb.HttpFilter_TypedConfig{
+			TypedConfig: testutils.MarshalAny(t, &v3xdsxdstypepb.TypedStruct{
+				TypeUrl: typeURL,
+				Value: &structpb.Struct{
+					Fields: map[string]*structpb.Value{
+						filterCfgPathFieldName:  {Kind: &structpb.Value_StringValue{StringValue: path}},
+						filterCfgErrorFieldName: {Kind: &structpb.Value_StringValue{StringValue: err}},
+					},
+				},
+			}),
+		},
+	}
+}
+
+// newStubServer returns a stub server that sends any filter config metadata
+// received as part of incoming RPCs to the provided channel.
+func newStubServer(metadataCh chan []string) *stubserver.StubServer {
+	return &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, _ *testpb.Empty) (*testpb.Empty, error) {
+			md, ok := metadata.FromIncomingContext(ctx)
+			if !ok {
+				return nil, status.Error(codes.InvalidArgument, "missing metadata")
+			}
+			select {
+			case metadataCh <- md.Get(filterCfgMetadataKey):
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+			return &testpb.Empty{}, nil
+		},
+		UnaryCallF: func(ctx context.Context, req *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+			md, ok := metadata.FromIncomingContext(ctx)
+			if !ok {
+				return nil, status.Error(codes.InvalidArgument, "missing metadata")
+			}
+			select {
+			case metadataCh <- md.Get(filterCfgMetadataKey):
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+			return &testpb.SimpleResponse{Payload: req.GetPayload()}, nil
+		},
+	}
+}
+
+// Tests HTTP filters with the xDS resolver. The test exercises various levels
+// of filter config overrides (base, virtual host-level, route-level and
+// cluster-level), and verifies that the correct config is applied for each RPC.
+func (s) TestXDSResolverHTTPFilters_AllOverrides(t *testing.T) {
+	// Override default WRR with a deterministic test version.
+	origNewWRR := rinternal.NewWRR
+	rinternal.NewWRR = testutils.NewTestWRR
+	defer func() { rinternal.NewWRR = origNewWRR }()
+
+	// Register a custom httpFilter builder for the test.
+	testFilterName := t.Name()
+	fb := &testHTTPFilterWithRPCMetadata{logger: t, typeURL: testFilterName}
+	httpfilter.Register(fb)
+	defer httpfilter.UnregisterForTesting(fb.typeURL)
+
+	// Spin up an xDS management server
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{AllowResourceSubset: true})
+	defer mgmtServer.Stop()
+
+	// Create an xDS resolver with bootstrap configuration pointing to the above
+	// management server.
+	nodeID := uuid.New().String()
+	bootstrapContents := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	if internal.NewXDSResolverWithConfigForTesting == nil {
+		t.Fatalf("internal.NewXDSResolverWithConfigForTesting is nil")
+	}
```

Comment:
nit: We can avoid this check and let the code panic if we don't expect to fail in most cases.